src/charm.py

```diff
@@ -113,85 +106,125 @@ def _on_jenkins_pebble_ready(self, event: PebbleReadyEvent) -> None:
 
-        self.unit.status = ActiveStatus("Checking for updates.")
+        The update will only take place if the current time is within the restart-time-range config
+        value.
+
+        Args:
+            container: The Jenkins workload container.
+
+        Returns:
+            The unit status of the charm after the operation.
+        """
+        original_status = self.unit.status.name
+        self.unit.status = ops.StatusBase.from_name(original_status, "Checking for updates.")
         try:
             latest_patch_version = jenkins.get_updatable_version(proxy=self.state.proxy_config)
         except jenkins.JenkinsUpdateError as exc:
             logger.error("Failed to get Jenkins updates, %s", exc)
-            self.unit.status = ActiveStatus("Failed to get Jenkins patch version.")
-            return
+            return ops.StatusBase.from_name(
+                original_status, "Failed to get Jenkins patch version."
+            )
 
         if not latest_patch_version:
-            self.unit.status = ActiveStatus()
-            return
+            return ops.StatusBase.from_name(original_status, "")
 
-        self.unit.status = MaintenanceStatus("Updating Jenkins.")
+        self.unit.status = ops.MaintenanceStatus("Updating Jenkins.")
         try:
-            jenkins.download_stable_war(self._jenkins_container, latest_patch_version)
+            jenkins.download_stable_war(container, latest_patch_version)
         except jenkins.JenkinsNetworkError as exc:
             logger.error("Failed to download Jenkins war. %s", exc)
-            self.unit.status = ActiveStatus("Failed to download executable.")
-            return
-
-        credentials = jenkins.get_admin_credentials(self._jenkins_container)
+            return ops.StatusBase.from_name(original_status, "Failed to download executable.")
         try:
-            jenkins.safe_restart(credentials)
+            jenkins.safe_restart(container)
             jenkins.wait_ready()
         except (jenkins.JenkinsError, TimeoutError) as exc:
             logger.error("Failed to safely restart Jenkins. %s", exc)
-            self.unit.status = BlockedStatus("Update restart failed.")
-            return
+            return ops.BlockedStatus("Update restart failed.")
 
         self.unit.set_workload_version(latest_patch_version)
-        self.unit.status = ActiveStatus()
+        return ops.ActiveStatus()
+
+    def _on_update_status(self, _: ops.UpdateStatusEvent) -> None:
+        """Handle update status event.
+
+        On Update status:
+        1. Remove plugins that are installed but are not allowed by plugins config value.
+        2. Update Jenkins patch version if available and is within restart-time-range config value.
+        """
+        container = self.unit.get_container(self.state.jenkins_service_name)
+        if not container.can_connect():
+            return
+
+        if self.state.restart_time_range and not self.state.restart_time_range.check_now():
```

Comment:
hmm, I'm slightly dissatisfied with the state exposing the `check_now` function, the state should be more like a data object rather than providing functionality. Could this just be in a separate function which takes the state as an argument and returns whether it is ok to restart?

Reply:
This has been updated in a separate module, thank you for the suggestion!